### PR TITLE
Feature Two Hand Manipulation + Bounding Box Normal Mode

### PR DESCRIPTION
**Pull request by cre8ivepark**
# Overview
mrtk_twohandmanipulation

This Pull Request is based on [TwoHandManipulatable script](https://github.com/johnppella/MixedRealityToolkit-Unity/blob/Feature_UX_BoundingBox_TwoHandManipulation/Assets/MixedRealityToolkit/InputModule/Scripts/Utilities/Interactions/TwoHandManipulatable.cs) made by Julia Schwarz(@julenka) and David Calabrese (Issue #1604)

This script allows for an object to be movable, scalable, and rotatable with one or two hands. You may also configure the script on only enable certain manipulations. The script works with both HoloLens' gesture input and immersive headset's motion controller input.

The example also includes Bounding Box Basic for the normal mode. In normal mode, Bounding Box Basic gives the visual boundary of the object when you interact with the object. You can find this behavior in the cliff house. (Grab and move, rotate, scale with two motion controllers)

mrtk_boundingboxnormalmode

# Demo Video
https://gfycat.com/ColossalEnlightenedAntipodesgreenparakeet

# Supported devices
- HoloLens 
- Immersive headset 

# Test Scene 
[TwoHandManipulationTest.unity](https://github.com/johnppella/MixedRealityToolkit-Unity/blob/Feature_UX_BoundingBox_TwoHandManipulation/Assets/MixedRealityToolkit-Examples/Input/Scenes/TwoHandManipulationTest.unity)

# Technical details 
You can find technical details on the README page: 
https://github.com/johnppella/MixedRealityToolkit-Unity/blob/Feature_UX_BoundingBox_TwoHandManipulation/Assets/MixedRealityToolkit-Examples/Input/Readme/README_TwoHandManipulationTest.md

# Known issues
- Adjust mode Bounding Box(handles) and App Bar will be added through different Pull Request.
- In the immersive headset, current input system only shows single pointer cursor. While holding an object with the first controller, pointing and selecting the object with the second controller will activate the manipulation behavior. Multi-pointer cursor visualization will be updated through different Pull Request.

This PR also includes the fix for some of the broken materials in Button prefabs used in the InteractableObject and ObjectCollection.

### CODE REVIEW

File: Assets/MixedRealityToolkit/InputModule/Scripts/Utilities/Interactions/TwoHandManipulatable.cs [hunk 1]
```diff
@@ -0,0 +1,631 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License. See LICENSE in the project root for license information.
+
+// #defines for features that are not yet implemented
+#define TODO_ROTATE_FACE_USER
```
Comment (StephenHodgson) [inline Assets/MixedRealityToolkit/InputModule/Scripts/Utilities/Interactions/TwoHandManipulatable.cs]: What's this doing exactly?

I'm not sure we should be defining our own symbols.

File: Assets/MixedRealityToolkit/InputModule/Scripts/Utilities/Interactions/TwoHandManipulatable.cs [hunk 2]
```diff
@@ -0,0 +1,631 @@
...
+    /// This script allows for an object to be movable, scalable, and rotatable with one or two hands. 
+    /// You may also configure the script on only enable certain manipulations. The script works with 
+    /// both HoloLens' gesture input and immersive headset's motion controller input.
+    /// See Assets/MixedRealityToolkit-Examples/Input/Readme/README_TwoHandManipulationTest.md
+    /// for instructions on how to use the script.
+    /// </summary>
+    public class TwoHandManipulatable : MonoBehaviour, IInputHandler, ISourceStateHandler
+    {
+        // Event that gets raised when user begins manipulating the object
+        public event Action StartedManipulating;
```
Comment (StephenHodgson) [inline Assets/MixedRealityToolkit/InputModule/Scripts/Utilities/Interactions/TwoHandManipulatable.cs]: Should we replace this with calls to the InputManager.RaiseManipulationStarted/Ended?

File: Assets/MixedRealityToolkit/InputModule/Scripts/Utilities/Interactions/TwoHandManipulatable.cs [hunk 3]
```diff
@@ -0,0 +1,631 @@
...
+        // Event that gets raised when the user ends manipulation
+        public event Action StoppedManipulating;
+        // Bounding Box visual on manipulation interaction
+        public GameObject GlobalBoundingBox;
```
Comment (StephenHodgson) [inline Assets/MixedRealityToolkit/InputModule/Scripts/Utilities/Interactions/TwoHandManipulatable.cs]: We should probably encapsulate this and serialize the private field

File: Assets/MixedRealityToolkit/InputModule/Scripts/Utilities/Interactions/TwoHandManipulatable.cs [hunk 4]
```diff
@@ -0,0 +1,631 @@
...
+        // Maps input id -> input source. Then obtain position of input source using currentInputSource.TryGetGripPosition(currentInputSourceId, out inputPosition);
+        private readonly Dictionary<uint, IInputSource> m_handsPressedInputSourceMap = new Dictionary<uint, IInputSource>();
+
+        private void Awake()
+        {
+            m_moveLogic = new MoveSphericalCoordsLogic();
+            m_rotateLogic = new HandlebarRotateLogic(ConstraintOnRotation);
+            m_scaleLogic = new ScaleLogic();
+
+            GlobalBoundingBox = GameObject.Find("BoundingBoxBasic");
```
Comment (StephenHodgson) [inline Assets/MixedRealityToolkit/InputModule/Scripts/Utilities/Interactions/TwoHandManipulatable.cs]: Where is this object located in relation to this GameObject the component is on?

File: Assets/MixedRealityToolkit/InputModule/Scripts/Utilities/Interactions/TwoHandManipulatable.cs [hunk 5]
```diff
@@ -0,0 +1,631 @@
...
+                }
+            }
+
+            if (currentState != State.Start)
+            {
+                UpdateStateMachine();
+            }
+        }
+
+        private void Start()
```
Comment (StephenHodgson) [inline Assets/MixedRealityToolkit/InputModule/Scripts/Utilities/Interactions/TwoHandManipulatable.cs]: please move the start method between awake and update.

File: Assets/MixedRealityToolkit/InputModule/Scripts/Utilities/Interactions/TwoHandManipulatable.cs [hunk 6]
```diff
@@ -0,0 +1,631 @@
...
+
+        private void OnOneHandMoveUpdated()
+        {
+            var targetPosition = m_moveLogic.Update(m_handsPressedLocationsMap.Values.First(), HostTransform.position);
+
+            HostTransform.position = targetPosition;
+
+#if !TODO_ROTATE_FACE_USER
+            // TODO: Rotate object to face user as needed.
+            TryRotateObjectToFaceUser();
```
Comment (StephenHodgson) [inline Assets/MixedRealityToolkit/InputModule/Scripts/Utilities/Interactions/TwoHandManipulatable.cs]: not sure I like using the `#if SYMBOL` here for a placeholder todo

File: Assets/MixedRealityToolkit/InputModule/Scripts/Utilities/Interactions/TwoHandManipulatable.cs [hunk 7]
```diff
@@ -0,0 +1,631 @@
...
+            }
+        }
+    }
+
+    /// <summary>
+    /// RotateLogic implements common logic for rotating holograms using a handlebar metaphor. 
+    /// When a manipulation starts, call Setup.
+    /// Call Update any time to update the move logic and get a new rotation for the object.
+    /// </summary>
+    public class HandlebarRotateLogic
```
Comment (StephenHodgson) [inline Assets/MixedRealityToolkit/InputModule/Scripts/Utilities/Interactions/TwoHandManipulatable.cs]: public classes should be in their own files.

File: Assets/MixedRealityToolkit/InputModule/Scripts/Utilities/Interactions/TwoHandManipulatable.cs [hunk 8]
```diff
@@ -0,0 +1,631 @@
...
+            m_draggingPosition = newPosition;
+
+
+            return m_draggingPosition;
+        }
+
+        /// <returns>A point that is below and just in front of the head.</returns>
+        public static Vector3 GetHandPivotPosition()
+        {
+            Vector3 pivot = Camera.main.transform.position + new Vector3(0, -0.2f, 0) - Camera.main.transform.forward * 0.2f; // a bit lower and behind
```
Comment (StephenHodgson) [inline Assets/MixedRealityToolkit/InputModule/Scripts/Utilities/Interactions/TwoHandManipulatable.cs]: use `CameraCache.Main`

File: Assets/MixedRealityToolkit/UX/Scripts/BoundingBoxes/BoundingBoxGizmoShell.cs [hunk 9]
```diff
@@ -9,7 +9,7 @@ namespace MixedRealityToolkit.UX.BoundingBoxes
     /// Draws a bounding box gizmo in the style of the hololens shell
     /// </summary>
     [ExecuteInEditMode]
-    public class BoundingBoxGizmoShell : BoundingBoxGizmo
+    public class BoundingBoxBasic : BoundingBoxGizmo
```
Comment (StephenHodgson) [inline Assets/MixedRealityToolkit/UX/Scripts/BoundingBoxes/BoundingBoxGizmoShell.cs]: We should probably use an obsolete marking in an empty class to tell devs what the new class name is.

**Comment by cre8ivepark** [comment]
@julenka Could you please take a look?

File: Assets/MixedRealityToolkit/InputModule/Scripts/Utilities/Interactions/TwoHandManipulatable.cs (hunk 1 above)
Comment (julenka) [reply Assets/MixedRealityToolkit/InputModule/Scripts/Utilities/Interactions/TwoHandManipulatable.cs]: It's just a way for me to keep track of TO DO items. I removed it.

File: Assets/MixedRealityToolkit/InputModule/Scripts/Utilities/Interactions/TwoHandManipulatable.cs (hunk 3 above)
Comment (julenka) [reply Assets/MixedRealityToolkit/InputModule/Scripts/Utilities/Interactions/TwoHandManipulatable.cs]: Resolved.

File: Assets/MixedRealityToolkit/InputModule/Scripts/Utilities/Interactions/TwoHandManipulatable.cs (hunk 5 above)
Comment (julenka) [reply Assets/MixedRealityToolkit/InputModule/Scripts/Utilities/Interactions/TwoHandManipulatable.cs]: Done.

File: Assets/MixedRealityToolkit/InputModule/Scripts/Utilities/Interactions/TwoHandManipulatable.cs (hunk 8 above)
Comment (julenka) [reply Assets/MixedRealityToolkit/InputModule/Scripts/Utilities/Interactions/TwoHandManipulatable.cs]: Out of curiosity, why is using `CameraCache.Main` better than Camera.main?

File: Assets/MixedRealityToolkit/InputModule/Scripts/Utilities/Interactions/TwoHandManipulatable.cs (hunk 2 above)
Comment (julenka) [reply Assets/MixedRealityToolkit/InputModule/Scripts/Utilities/Interactions/TwoHandManipulatable.cs]: Hm, based on my reading of `CustomInputSource.cs` it seems like manipulation events are fired whenever a motion controller is moved and a button is down, regardless of whether something is actually being manipulated. Further, the events contain information about the controller id attached to the manipulation. In contrast the `StartedManipulating` and `StoppedManipulating` events are specific to this object--the event communicates "a specific object began being manipulated". Additionally, an object may be manipulated by more than one input source, which would not fit with the current structure of the `ManipulationEventData`. For these two reasons, I think it's better to keep the events as separate.

File: Assets/MixedRealityToolkit/InputModule/Scripts/Utilities/Interactions/TwoHandManipulatable.cs (hunk 4 above)
Comment (julenka) [reply Assets/MixedRealityToolkit/InputModule/Scripts/Utilities/Interactions/TwoHandManipulatable.cs]: @cre8ivepark should comment more here, but my understanding is that this 'BoundingBoxBasic" UI element can be located anywhere in the scene. Actually after reading the code, we may want to just have a separate component "ShowBoundingBoxWhenManipulating" that would listen to the manipulation started and stopped events, and show/hide the bounding box. That would separate the concerns of two handed manipulation (which just manipulates objects) and the rendering of feedback (showing the bounding box). I'll see if I can do this over the weekend, or Yoon/somebody else please feel free to make the change.

**Comment by julenka** [comment]
Made a few changes tonight but need to go to sleep, will try to pick this up over the weekend since I'll be busy Friday night.

File: Assets/MixedRealityToolkit/InputModule/Scripts/Utilities/Interactions/TwoHandManipulatable.cs (hunk 2 above)
Comment (julenka) [reply Assets/MixedRealityToolkit/InputModule/Scripts/Utilities/Interactions/TwoHandManipulatable.cs]: The pattern I emulated was the `StartedDragging` and `StoppedDragging` events in `HandDraggable.cs`

Comment (StephenHodgson) [reply Assets/MixedRealityToolkit/InputModule/Scripts/Utilities/Interactions/TwoHandManipulatable.cs]: It doesn't matter what the input source is, we can take advantage of these events here.

File: Assets/MixedRealityToolkit/InputModule/Scripts/Utilities/Interactions/TwoHandManipulatable.cs (hunk 4 above)
Comment (StephenHodgson) [reply Assets/MixedRealityToolkit/InputModule/Scripts/Utilities/Interactions/TwoHandManipulatable.cs]: Just wondering if we can assign the GameObject in the inspector and maybe provide a way to search if it's unset.

File: Assets/MixedRealityToolkit/InputModule/Scripts/Utilities/Interactions/TwoHandManipulatable.cs (hunk 8 above)
Comment (StephenHodgson) [reply Assets/MixedRealityToolkit/InputModule/Scripts/Utilities/Interactions/TwoHandManipulatable.cs]: Cause tag lookups are expensive and this eliminates the issue.

File: Assets/MixedRealityToolkit/InputModule/Scripts/Utilities/Interactions/TwoHandManipulatable.cs (hunk 4 above)
Comment (cre8ivepark) [reply Assets/MixedRealityToolkit/InputModule/Scripts/Utilities/Interactions/TwoHandManipulatable.cs]: @johnppella

File: Assets/MixedRealityToolkit/UX/Scripts/BoundingBoxes/BoundingBoxGizmoShell.cs (hunk 9 above)
Comment (johnppella) [reply Assets/MixedRealityToolkit/UX/Scripts/BoundingBoxes/BoundingBoxGizmoShell.cs]: i changed the type of the variable BoundingBoxGizmo. The name of the class didn't change.

File: Assets/MixedRealityToolkit/InputModule/Scripts/Utilities/Interactions/TwoHandManipulatable.cs (hunk 4 above)
Comment (johnppella) [reply Assets/MixedRealityToolkit/InputModule/Scripts/Utilities/Interactions/TwoHandManipulatable.cs]: changed search for GlobalBoundingBox in scene to Instatiation of BoundingBox prefab.

File: Assets/MixedRealityToolkit/InputModule/Scripts/Utilities/Interactions/TwoHandManipulatable.cs (hunk 2 above)
Comment (julenka) [reply Assets/MixedRealityToolkit/InputModule/Scripts/Utilities/Interactions/TwoHandManipulatable.cs]: What about this example: There's some code that listens for ManipulationStarted and ManipulationEnded, assuming that this event gets fired for each controller when the controller starts dragging. Then the person adds a two handed manip component to their scene and is surprised to see a new ManipulationStarted event firing whenever the two hand manipulatable starts moving, in addition to the two ManipulationStarted events from each controller? 

Additionally, wouldn't this break the pattern that HandDraggable already has? A developer might be confused about why OneHandDraggable exposes a StartedDragging event, but TwoHandManipulatable does not, and instead for that one the dev needs to implement IManipulatable, or register events listeners in their code.

File: Assets/MixedRealityToolkit/InputModule/Scripts/Utilities/Interactions/TwoHandManipulatable.cs (hunk 7 above)
Comment (julenka) [reply Assets/MixedRealityToolkit/InputModule/Scripts/Utilities/Interactions/TwoHandManipulatable.cs]: Fixed.

**Comment by julenka** [comment]
Hi @StephenHodgson I've made a few more changes based on your comments, I still am not sure about using the `InputManager.RaiseManipulationStarted` event though. Things looks good in the test scene when running in editor, but I have not yet tested on device. Hope to get to that later this week or this coming weekend.

File: Assets/MixedRealityToolkit/UX/Scripts/BoundingBoxes/BoundingBoxGizmoShell.cs (hunk 9 above)
Comment (cre8ivepark) [reply Assets/MixedRealityToolkit/UX/Scripts/BoundingBoxes/BoundingBoxGizmoShell.cs]: This class name change was unnecessary. Reverted this change.

File: Assets/MixedRealityToolkit/InputModule/Scripts/Utilities/Interactions/TwoHandManipulatable.cs (hunk 2 above)
Comment (julenka) [reply Assets/MixedRealityToolkit/InputModule/Scripts/Utilities/Interactions/TwoHandManipulatable.cs]: Okay, I'll look at that this weekend.

**Comment by HBJKZLH** [comment]
Untiy [Universal Windows Platform] Platform does not support.

The Socket code is written and the Thread code is written and the FileStream code is written, you know what?



I used the network communication section to find that this part of the code could not be transferred to the [Universal Windows Platform].

After I tested Unity 2017.2.0f3 (64-bit) version,Unity 2017.3.1f1 (64-bit) version,Unity 2017.2.1f1 (64-bit) version.



They don't support it.



Do you have any feedback on untiy?









When I wrote socket and IO and Thread related code in the program, untiy packaging times was wrong,
The socket.Connect function is not supported.
The socket.Receive function is not supported, and the 
socket.Send function is not supported.
The filestream.close function is not supported,
New Thread(recv) is not supported,
Does not support Thread. IsBackground,
Does not support Thread. To Start,




But if I just write in the code.

The Socket client = null;

Thread the Thread = null;

FileStream filedata = null;



This way, the Universal Windows Platform is passed.

I think it should be the flaw in the Platform library of untiy [Universal Windows Platform].


Please help me to feedback this question,








------------------ 原始邮件 ------------------
发件人: "Stephen Hodgson"<notifications@github.com>;
发送时间: 2018年3月2日(星期五) 上午7:17
收件人: "Microsoft/MixedRealityToolkit-Unity"<MixedRealityToolkit-Unity@noreply.github.com>;
抄送: "Subscribed"<subscribed@noreply.github.com>; 
主题: Re: [Microsoft/MixedRealityToolkit-Unity] Feature Two HandManipulation + Bounding Box Normal Mode (#1727)




@StephenHodgson approved this pull request.

**Comment by HBJKZLH** [comment]
Untiy [Universal Windows Platform] Platform does not support.

The Socket code is written and the Thread code is written and the FileStream code is written, you know what?



I used the network communication section to find that this part of the code could not be transferred to the [Universal Windows Platform].

After I tested Unity 2017.2.0f3 (64-bit) version,Unity 2017.3.1f1 (64-bit) version,Unity 2017.2.1f1 (64-bit) version.



They don't support it.



Do you have any feedback on untiy?









When I wrote socket and IO and Thread related code in the program, untiy packaging times was wrong,
The socket.Connect function is not supported.
The socket.Receive function is not supported, and the 
socket.Send function is not supported.
The filestream.close function is not supported,
New Thread(recv) is not supported,
Does not support Thread. IsBackground,
Does not support Thread. To Start,




But if I just write in the code.

The Socket client = null;

Thread the Thread = null;

FileStream filedata = null;



This way, the Universal Windows Platform is passed.

I think it should be the flaw in the Platform library of untiy [Universal Windows Platform].


Please help me to feedback this question,








------------------ 原始邮件 ------------------
发件人: "Julia Schwarz"<notifications@github.com>;
发送时间: 2018年3月2日(星期五) 凌晨2:49
收件人: "Microsoft/MixedRealityToolkit-Unity"<MixedRealityToolkit-Unity@noreply.github.com>;
抄送: "Subscribed"<subscribed@noreply.github.com>; 
主题: Re: [Microsoft/MixedRealityToolkit-Unity] Feature Two HandManipulation + Bounding Box Normal Mode (#1727)




@julenka commented on this pull request.
 
 
In Assets/MixedRealityToolkit/InputModule/Scripts/Utilities/Interactions/TwoHandManipulatable.cs:
 > +using UnityEngine.Assertions; + + +namespace MixedRealityToolkit.InputModule.Utilities.Interations +{ +    /// <summary> +    /// This script allows for an object to be movable, scalable, and rotatable with one or two hands.  +    /// You may also configure the script on only enable certain manipulations. The script works with  +    /// both HoloLens' gesture input and immersive headset's motion controller input. +    /// See Assets/MixedRealityToolkit-Examples/Input/Readme/README_TwoHandManipulationTest.md +    /// for instructions on how to use the script. +    /// </summary> +    public class TwoHandManipulatable : MonoBehaviour, IInputHandler, ISourceStateHandler +    { +        // Event that gets raised when user begins manipulating the object +        public event Action StartedManipulating;  
Okay, I'll look at that this weekend.

**Comment by cre8ivepark** [comment]
@julenka @StephenHodgson I think we can merge this and consider adding a new event as an improvement. Thanks!

**Comment by cre8ivepark** [comment]
I'll merge this update to our Bounding Box and App Bar branch and submit PR.

**Pull request merged**